src/tree.rs

```diff
@@ -0,0 +1,662 @@
+                result_meta
+                    .insert("offset", (offset as i64).into())
+                    .insert("size", (res.len() as i64).into());
+                Ok(RpcValue::new(res.into(), Some(result_meta)))
+            }
+            METH_READ_COMPRESSED => {
+                let read_params: ReadParams = rq
+                    .param()
+                    .unwrap_or_default()
+                    .try_into()
+                    .map_err(|msg| RpcError::new(RpcErrorCode::InvalidParam, msg))?;
+                let offset = read_params.offset.unwrap_or(0);
+                let size = read_params.size.unwrap_or(0);
+                let (res, bytes_read) = compress_file(path, offset, size)
+                    .await
+                    .map_err(rpc_error_filesystem)?;
+                let mut result_meta = shvproto::MetaMap::new();
+                result_meta
+                    .insert("offset", (offset as i64).into())
+                    .insert("size", (bytes_read as i64).into());
+                Ok(RpcValue::new(res.into(), Some(result_meta)))
+            }
+            _ => Err(rpc_error_unknown_method(method)),
+        }
+    } else {
+        Err(rpc_error_unknown_method(method))
+    }
+}
+
+#[derive(Default)]
+struct ReadParams {
+    offset: Option<u64>,
+    size: Option<u64>,
+}
+
+impl TryFrom<&RpcValue> for ReadParams {
+    type Error = String;
+
+    fn try_from(value: &RpcValue) -> Result<Self, Self::Error> {
+        if value.is_null() {
+            return Ok(Self::default());
+        }
+
+        let map: shvproto::rpcvalue::Map = value.try_into()?;
+
+        let parse_param = |param_name: &str| -> Result<Option<u64>, String> {
+            match map.get(param_name) {
+                None => Ok(None),
+                Some(val) => {
+                    i64::try_from(val)
+                        .map_err(|e| e.to_string())
+                        .and_then(|v| u64::try_from(v)
+                            .map_err(|e| e.to_string())
+                        )
+                        .map_err(|e| format!("Error parsing `{param_name}` parameter: {e}"))
+                        .map(Some)
+                }
+            }
+        };
+
+        let offset = parse_param("offset")?;
+        let size = parse_param("size")?;
+
+        Ok(Self { offset, size })
+    }
+}
+
+
+async fn with_file_reader<F, Fut, R>(
+    path: impl AsRef<str>,
+    offset: u64,
+    size: u64,
+    process: F,
+) -> tokio::io::Result<R>
+where
+    F: FnOnce(Box<dyn AsyncBufRead + Unpin + Send>) -> Fut + Send,
+    Fut: Future<Output = tokio::io::Result<R>> + Send,
+{
+    const MAX_READ_SIZE: u64 = 10 * (1 << 20);
+    let mut file = tokio::fs::File::open(path.as_ref()).await?;
+    file.seek(std::io::SeekFrom::Start(offset)).await?;
+    let file_size = file.metadata().await?.len();
+    let size = if size == 0 { file_size } else { size }.min(file_size).min(MAX_READ_SIZE);
```

Comment:
We should standardize the behavior if argument `size == 0`. Current doc only specifies a specific scenario: https://silicon-heaven.github.io/shv-doc/rpcmethods/file.html#read

Reply:
The function is refactored to accept an Option parameter, which defaults to `file_size` if None. No special handling of `size == 0` is implemented.